tfhe/src/integer/server_key/comparator.rs

```diff
@@ -313,36 +372,142 @@ impl<'a> Comparator<'a> {
             comparisons
         };
 
-        let mut comparisons_2 = Vec::with_capacity(comparisons.len() / 2);
-        while comparisons.len() != 1 {
-            comparisons
-                .par_chunks_exact(2)
-                .map(|chunk| {
-                    let (low, high) = (&chunk[0], &chunk[1]);
-                    let mut high = high.clone();
-
-                    // We don't use pack_block_assign as the offset '4' does not depend on params
-                    self.server_key
-                        .key
-                        .unchecked_scalar_mul_assign(&mut high, 4);
-                    self.server_key.key.unchecked_add_assign(&mut high, low);
+        self.reduce_signs_parallelized(comparisons)
+    }
 
-                    self.server_key
-                        .key
-                        .apply_lookup_table_assign(&mut high, &self.selection_accumulator);
-                    high
-                })
-                .collect_into_vec(&mut comparisons_2);
+    fn unchecked_scalar_block_slice_sign_parallelized<'b, PBSOrder>(
+        &self,
+        lhs_blocks: &'b [CiphertextBase<PBSOrder>],
+        scalar_blocks: &[u8],
+    ) -> Vec<crate::shortint::CiphertextBase<PBSOrder>>
+    where
+        PBSOrder: PBSOrderMarker,
+        &'b [CiphertextBase<PBSOrder>]:
+            rayon::iter::IntoParallelIterator<Item = &'b CiphertextBase<PBSOrder>>,
+    {
+        assert_eq!(lhs_blocks.len(), scalar_blocks.len());
+        let num_blocks = lhs_blocks.len();
+        let num_blocks_halved = (num_blocks / 2) + (num_blocks % 2);
+
+        let message_modulus = self.server_key.key.message_modulus.0;
+        let mut signs = Vec::with_capacity(num_blocks_halved);
+        lhs_blocks
+            .par_chunks(2)
+            .zip(scalar_blocks.par_chunks(2))
+            .map(|(lhs_chunk, scalar_chunk)| {
+                let packed_scalar = scalar_chunk[0]
+                    + (scalar_chunk.get(1).copied().unwrap_or(0) * message_modulus as u8);
+                let mut packed_lhs = self.pack_block_chunk(lhs_chunk);
+                self.scalar_compare_block_assign(&mut packed_lhs, packed_scalar);
+                packed_lhs
+            })
+            .collect_into_vec(&mut signs);
 
-            if (comparisons.len() % 2) == 1 {
-                comparisons_2.push(comparisons[comparisons.len() - 1].clone());
-            }
+        signs
```

Comment:
here you don't manage chunks exact and skip the clone any reason why compared to another function doing something similar ?

Reply:
The old code can be updated to use par_chunks.

In this pr I modified the `pack_block_chunk` to handle the case when `chunk.len() == 1`  which is why here I use par_chunks.